File: kotlin/algorithms/other/FloydsCycle.kt
package algorithms.other

import java.util.LinkedList

data class Node(val value: Int, var next: Node? = null)

fun get(list: LinkedList<Node>, index: Int): Node? =
    list.getOrNull(index)

fun floydsCycle(list: LinkedList<Node>): Boolean {
    var slowIndex = 0
    var fastIndex = 0

    if (list.isEmpty()) return false

    while (get(list, slowIndex) != null
        && get(list, fastIndex) != null
        && get(list, fastIndex)?.next != null
    ) {
        val next = get(list, slowIndex)!!.next!!.value
        slowIndex = next
        fastIndex = get(list, next)!!.next!!.value

        if (get(list, slowIndex) == get(list, fastIndex)) {
            break
        }
    }

    if (get(list, slowIndex) != get(list, fastIndex)) {
        return false
    }

    // reset slow pointer to head
    // and traverse again
    slowIndex = 0
    while (get(list, slowIndex) != get(list, fastIndex)) {
        slowIndex++
        slowIndex++
    }

    return slowIndex == fastIndex;
}

File: dart/algorithms/other/floyds_cycle.dart
import 'dart:collection';

final class EntryItem extends LinkedListEntry<EntryItem> {
  EntryItem(this.id);

  final int id;
}

bool floydsCycle(LinkedList<EntryItem> list) {
  var slowPointer = list.firstOrNull;
  var fastPointer = list.firstOrNull;

  while (
      slowPointer != null && fastPointer != null && fastPointer.next != null) {
    final next = slowPointer.next;
    slowPointer = next;
    fastPointer = next?.next;

    if (slowPointer?.id == fastPointer?.id) {
      break;
    }
  }

  if (slowPointer?.id != fastPointer?.id) {
    return false;
  }

  // reset slow pointer to head
  // and traverse again
  slowPointer = list.firstOrNull;
  while (slowPointer?.id != fastPointer?.id) {
    final next = slowPointer!.next;
    slowPointer = next;
    fastPointer = next?.next;
  }

  return slowPointer?.id == fastPointer?.id;
}
